proofs/erc20.md

```diff
@@ -0,0 +1,1030 @@
+
+callData: #abiCallData("balanceOf", #address(OWNER))
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(BAL, 32) ]
+      _:Map
+    )
+
+gas: {gascap} => _
+
+log: _
+
+refund: _
+```
+
+`<storage>` cell specifies that the value of `balances[OWNER]` is `BAL`, which will be returned as described in `<localMem>` cell.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(balances)}, OWNER) |-> BAL
+    _:Map
+
+requires:
+    andBool 0 <=Int OWNER     andBool OWNER     <Int (2 ^Int 160)
+    andBool 0 <=Int BAL       andBool BAL       <Int (2 ^Int 256)
+
+comments:
+```
+
+allowance
+---------
+
+The specification of `allowance` is similar to that of `totalSupply` as well.
+
+```{.k .erc20-spec .hkg-spec}
+[allowance]
+
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("allowance", #address(OWNER), #address(SPENDER))
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(ALLOWANCE, 32) ]
+      _:Map
+    )
+
+gas: {gascap} => _
+
+log: _
+
+refund: _
+```
+
+`<storage>` cell specifies that the value of `allowances[OWNER][SPENDER]` is `ALLOWANCE`, which will be returned as described in `<localMem>` cell.
+
+```{.k .erc20-spec .hkg-spec}
+storage:
+    #hashedLocation({lang}, {pos(allowances)}, OWNER SPENDER) |-> ALLOWANCE
+    _:Map
+
+requires:
+    andBool 0 <=Int OWNER     andBool OWNER     <Int (2 ^Int 160)
+    andBool 0 <=Int SPENDER   andBool SPENDER   <Int (2 ^Int 160)
+    andBool 0 <=Int ALLOWANCE andBool ALLOWANCE <Int (2 ^Int 256)
+
+comments:
+```
+
+
+approve
+-------
+
+```{.k .erc20-spec .hkg-spec}
+[approve]
+
+k: #execute => (RETURN RET_ADDR:Int 32 ~> _)
+
+callData: #abiCallData("approve", #address(SPENDER), #uint256(VALUE))
+
+localMem:
+      .Map
+    =>
+    (
+      .Map[ RET_ADDR := #asByteStackInWidth(1, 32) ]
+      _:Map
+    )
+
+gas: {gascap} => _
+```
+
+`<log>` cell specifies that an event log is generated during the execution.
+
+The log message contains the account ID of the current contract, the hash of the signature of event Approval, the account ID who calls this contract, the SPENDER account ID, and the approved VALUE. `#generateSignature` is a syntactic sugar, used to define the ABI call abstraction as well, that generates the event signature to be hashed.
+
+```{.k .erc20-spec .hkg-spec}
+log:
+      _:List
+    (
+      .List
+    =>
+      ListItem({ ACCT_ID
+               | #parseHexWord(Keccak256(#generateSignature("Approval", #address(CALLER_ID), #address(SPENDER), #uint256(VALUE))))
+               : CALLER_ID
+               : SPENDER
+               : .WordStack
+               | #asByteStackInWidth(VALUE, 32)
+               })
+    )
+```
+
+`<refund>` cell specifies that a refund may be issued. This function will refund a gas if the allowance value is 0 while the previous value is greater than 0, re-claiming the corresponding entry of the storage.
```

Comment:
The refund is part of EVM semantics. I think it would be great if IELE have a nice gas model *without* this, which I think could be yet another advantage over EVM.

Reply:
@daejunpark can you please open this discussion in the #iele channel on Slack?  I agree.  One question here is what is the right way to reclaim space.  In EVM, you reclaim when you write 0 in it, which results I guess in a refund, but is that such a big deal to be worth a refund?

Reply:
I would also add a sentence where the refund _ => _ appears for the first time, saying that this is something EVM-specific and there is no way to go around it.